From a8af6f5e11003e8d256c27cdc7d6d4b52167d760 Mon Sep 17 00:00:00 2001
From: Eric Lapuyade <eric.lapuyade@linux.intel.com>
Date: Mon, 29 Apr 2013 17:47:42 +0200
Subject: [PATCH 03/44] NFC: HCI: Implement fw_upload ops

This is a simple forward to the HCI driver. When driver is done with the
operation, it shall directly notify NFC Core by calling
nfc_fw_upload_done().

Signed-off-by: Eric Lapuyade <eric.lapuyade@intel.com>
Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>
(cherry picked from commit f875cdb417fdbbd81f97538175adaed63bb03974)
---
 include/net/nfc/hci.h |  1 +
 net/nfc/hci/core.c    | 11 +++++++++++
 2 files changed, 12 insertions(+)

diff --git a/include/net/nfc/hci.h b/include/net/nfc/hci.h
index b87a169..14faf2d 100644
--- a/include/net/nfc/hci.h
+++ b/include/net/nfc/hci.h
@@ -61,6 +61,7 @@ struct nfc_hci_ops {
 			      struct sk_buff *skb);
 	int (*enable_se)(struct nfc_dev *dev, u32 secure_element);
 	int (*disable_se)(struct nfc_dev *dev, u32 secure_element);
+	int (*fw_upload)(struct nfc_hci_dev *hdev, const char *firmware_name);
 };
 
 /* Pipes */
diff --git a/net/nfc/hci/core.c b/net/nfc/hci/core.c
index 91020b2..b7e4dac 100644
--- a/net/nfc/hci/core.c
+++ b/net/nfc/hci/core.c
@@ -779,6 +779,16 @@ static void nfc_hci_recv_from_llc(struct nfc_hci_dev *hdev, struct sk_buff *skb)
 	}
 }
 
+static int hci_fw_upload(struct nfc_dev *nfc_dev, const char *firmware_name)
+{
+	struct nfc_hci_dev *hdev = nfc_get_drvdata(nfc_dev);
+
+	if (hdev->ops->fw_upload)
+		return hdev->ops->fw_upload(hdev, firmware_name);
+
+	return -ENOTSUPP;
+}
+
 static struct nfc_ops hci_nfc_ops = {
 	.dev_up = hci_dev_up,
 	.dev_down = hci_dev_down,
@@ -791,6 +801,7 @@ static struct nfc_ops hci_nfc_ops = {
 	.im_transceive = hci_transceive,
 	.tm_send = hci_tm_send,
 	.check_presence = hci_check_presence,
+	.fw_upload = hci_fw_upload,
 };
 
 struct nfc_hci_dev *nfc_hci_allocate_device(struct nfc_hci_ops *ops,
-- 
1.8.4.93.g57e4c17

